packages/components/src/DateTimePickers/pickers/DatePicker/DatePicker.component.js

```diff
@@ -1,83 +1,166 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 import classNames from 'classnames';
+import memoize from 'lodash/memoize';
+import { chunk } from 'lodash';
+import format from 'date-fns/format';
+import endOfMonth from 'date-fns/end_of_month';
+import startOfWeek from 'date-fns/start_of_week';
+import isToday from 'date-fns/is_today';
+import getDate from 'date-fns/get_date';
+import setDay from 'date-fns/set_day';
+import getMonth from 'date-fns/get_month';
+import isSameDay from 'date-fns/is_same_day';
+import differenceInCalendarWeeks from 'date-fns/difference_in_calendar_weeks';
+
+import addDays from 'date-fns/add_days';
 import theme from './DatePicker.scss';
 import DayPickerAction from './DayPickerAction';
 
-function DatePicker(props) {
-	const dayNames = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];
-	const days = (new Array(7))
-					.fill(0)
-					.map((_, i) => i + 1);
+const FIRST_DAY_OF_WEEK = 1;
+const BASE_DATE = new Date(0);
 
-	const weeks = (new Array(4))
+function buildDayNames(firstDayOfweek) {
+	return (new Array(7))
 		.fill(0)
-		.map((_, i) =>
-			days.map(x => ({
-				number: x + (i * 7),
-			}))
-		);
+		.map((_, i) => (i + firstDayOfweek) % 7)
+		.map(dayOfWeek => setDay(BASE_DATE, dayOfWeek))
+		.map(headerDate => format(headerDate, 'dddd'));
+}
+
+const getDayNames = memoize(buildDayNames);
+
+class DatePicker extends React.Component {
+
+	static buildWeeks(year, monthIndex, firstDayOfWeek) {
+		const firstDateOfMonth = new Date(year, monthIndex);
+		const firstDateOfCalendar = startOfWeek(firstDateOfMonth, {
+			weekStartsOn: firstDayOfWeek,
+		});
 
-	const selectedDay = 12;
-	const currentDay = 20;
-	const disabledDays = [6, 15];
+		const lastDateOfMonth = endOfMonth(firstDateOfMonth);
+		const diffWeeks = differenceInCalendarWeeks(lastDateOfMonth, firstDateOfCalendar, {
+			weekStartsOn: firstDayOfWeek,
+		});
+		const nbWeeksToRender = diffWeeks + 1;
 
-	function isSelectedDay(n) {
-		return selectedDay === n;
+		const dates = (new Array(7 * nbWeeksToRender))
+						.fill(0)
+						.map((_, i) => addDays(firstDateOfCalendar, i));
+
+		return chunk(dates, 7);
 	}
 
-	function isCurrentDay(n) {
-		return currentDay === n;
+	constructor(props) {
+		super(props);
+
+		this.getWeeks = memoize(DatePicker.buildWeeks, (year, monthIndex, firstDayOfWeek) => `${year}-${monthIndex}|${firstDayOfWeek}`);
 	}
 
-	function isDisabledDay(n) {
-		return disabledDays.includes(n);
+	isSelectedDate(date) {
+		return this.props.selectedDate !== undefined
+			&& isSameDay(this.props.selectedDate, date);
 	}
 
-	return (
-		<div className={theme.container}>
-			<div className={classNames(theme['calendar-row'], theme['calendar-header-row'])}>
-				{dayNames.map((dayName, i) =>
-					<abbr
-						className={theme['calendar-item']}
-						key={i}
-						title={dayName}
-					>
-						{dayName.charAt(0)}
-					</abbr>
-				)}
-			</div>
+	isDisabledDate(dateToCheck) {
```

Comment:
This seems is quite a complicated api + code to handle that.

You should accept an array of functions, or an array of functions or dates. 
But not an array of arrays :/
Moreover, the condition of the top level array is an OR, and the conditions between each element in the nested arrays are OR too. So let's flatten.

BTW, What is the use case of having a specific date disabled ?

Reply:
i suppose disabled for :
- date min/max allowed
- range white or black list
- specific dates white or black list

I don't know the real constraints

But an array of array have some sens.
The first one is just for a set of rules, it can actually be one single date, a list of single date, and later eventually a range, a min/max, a custom function, etc. So the second array is just a specific rule.

But yes we don't know the needs actually, moreover, the month and year picker will need to have some disabled actions too. So maybe handle that at this level is too low, maybe handle that at a higher component level and have just function as prop to check dates (months and years) is fine.

Reply:
❓

Reply:
You can do that with an array of (Date || function)
If you have a function, you run it and take the result as boolean.
If you have a Date, you test `isSameDay`

No need to take an array of arrays.
* date min/max allowed: function
* range white or black list: function 
* specific dates black list: Dates
* specific dates white list: function

Reply:
✔